Scaleh

kdraeder opened the pull request:
## Description:
Define scale height with the straightforward, conventional definition.
Define "unnormalized" scale height consistent with that.
Define vertical threshholds using no_normalization_of_scale_heights, instead of hard coding .false.

Renamed variable higher_is_smaller to higher_is_decreasing.
This is (still) ambiguous, but keeps part of the old name and is less cumbersome than something unambigous
like gradient_with_height_is_negative.

Added debugging code (default is off) to above_ramp_start to help with verifying that it's working.

Applied nint to the reals returned by query_location.  This may not have been broken, but it seemed worth doing
while I was debugging.

~~Fix the order of dimensions in model_interpolate: call quad_lat_lon_evaluate(...quad_vals) to match the definition.  This isn't related to the vertical coordinate issues, but is a trivial fix of something 
that keeps surprising me.~~ HK edit: this was already fixed on main.

### Fixes issue
#298 

### Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] Documentation update

### Documentation changes needed?
<!-- Put an `x` in all the boxes that apply: -->
- [x] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly.

### Tests
Please describe any tests you ran to verify your changes.
I ran an assimilation using:

- just radiosonde U winds, 
- no_normalization_of_scale_heights = .true.
- model_damping_ends_at_level = 7 (a level below the highest U obs)

The program entered the appropriate sections of above_ramp_start for the heights of the observations.
There were no increments at the levels above model_damping_ends_at_level.

## Checklist for merging

- [ ] Updated changelog entry
- [ ] Documentation updated
- [ ] Version tag 

## Testing Datasets

- [ ] Dataset needed for testing available upon request
- [ ] Dataset download instructions included
- [ ] No dataset needed

hkershaw-brown commented:
Hello, where did you branch this from and where do you want it to go? main and main?
At the moment, it is against Manhattan, so there are 1182 files changed, which seems a little excessive.

hkershaw-brown commented:
You can edit the pull request to change the base branch to main.
![Screen Shot 2021-10-12 at 9 32 03 AM](https://user-images.githubusercontent.com/20047007/136986083-883d052d-0375-45b3-a991-162844167d70.png)

Your scaleh is a couple of months behind main. 
![Screen Shot 2021-10-12 at 9 31 05 AM](https://user-images.githubusercontent.com/20047007/136986119-1ac26b69-a4cc-4679-9792-2431a183e7f0.png)

You can make your branch update with main:

```
git checkout main
git pull
git checkout scaleh
git merge main scaleh
```

nancycollins commented:
we should have a consistent dart-wide definition of scale height, not only in the cam interface code.  being able to specify the vertical localization in units of scale height is an option for any model that uses the threed_sphere locations code.  specifically i know the wrf model_mod has lots of code that deals with scale height.  if these changes are consistent with how wrf or any other model_mod that includes scale height conversion code, then i'm all for it.  otherwise we need to be careful to not make inconsistent definitions or non-backwards compatible changes.  for example - do we know what vertical localization choices the waccm folks use?  if scale height, does this change their results?

kdraeder commented:
I changed the base and updated my local main and scaleh branch.

Nancy, good point.  I'll try to dig that stuff up.  I've never looked in those models,
so I don't know how long that will take.

kdraeder commented:
> kevin - when you tested this new code, what did you use for the scale height normalization factor in the locations module? would it be around the log of whatever the pressure normalization factor is or different? it would be interesting to see the differences between setting the vertical localization to pressure and then to scale height. do you have any plots of those?
I used 
vert_normalization_scale_height = 1.5
We can't compare the 2 using a simple log(p) because there's not a linear relationship between them.
They're used to decide whether a vertical distance between 2 points is large or small.
They're both distances, but they represent space differently and have different units.
Near the surface points that are 20000 Pa apart are much closer (geometrically) 
than if 1 of the points were in the thermosphere and separated by 20000 Pa.  
Scaled height preserves the measure of closeness with altitude, which is why we use it
for high topped models.
So 1.5 scaled heights covers a lot of pressure difference near the surface,
but only a little pressure difference in the upper atmosphere.

vert_normalization_pressure = 20000  
Calculating log(20000) is using 1 Pa as a reference pressure and yields the scaled height difference 
between 20000 and 1 hPa = 9.9 scaled heights.
The same 20000 Pa distance _below_ 20000 Pa translates into log(40000/20000) = 0.69 scaled heights.

We did comparisons of using the 2 different vertical coordinates (pressure and scaled height) years ago.
They perform similarly in the lower atmosphere, but scaled height is much better in the upper atmosphere,
which was the motivation for implementing it.
> 
> i didn't see where you added a global variable for the scale height cutoff and removed the local variable in the init routine, but you'd catch that with testing because it wouldn't work otherwise.
I didn't add it to this commit because it's more closely related to another issue that I'm planning to open.

nancycollins commented on models/cam-fv/model_mod.f90:
```diff
@@ -3584,7 +3584,8 @@ subroutine init_discard_high_obs()
 ! incoming observations which have vertical units of scale height.
 ! so far we have localized in scale height but never had obs
```
this comment is misleading and should be updated at some point, since you're moving the call to 'obs_too_high()' to the get_close routines.  the call in model_interpolate() hasn't seen incoming obs with scale height as vertical, but that's getting removed.  at get_close() time the obs are converted to the vertical localization type, so definitely yes there will be obs that have this type.

nancycollins commented on models/cam-fv/model_mod.f90:
```diff
@@ -3691,6 +3699,9 @@ function above_ramp_start(test_value, gc, obs_type, ramp_end, total_dist, extra_
+! ? Are those values used to set the maxdist in the gc?
 vert_localize_dist = get_maxdist(gc, obs_type)
 if (.false. .and. onetime) then
    print *, 'vert_localize_dist = ', vert_localize_dist
```
this debugging code shouldn't have been in the committed version either.  that's my fault.

hkershaw-brown commented:
I'm closing this since I think we've ended up reviewing code at bit too early. 

Lets step back from reviewing the code, and start with:
1. Documentation of what the code is supposed to do.
2. A link to restart files + obs_seq that lets someone reproduce the issue (on Cheyenne or wherever).  
3. If this is going to change people's results, who does it affect. e.g does this affect everyone who localizes in SCALEHEIGHT? All WACCM users? Is this a non-bitwise change for CAM users?

hkershaw-brown commented:
moved scaleh branch to https://github.com/hkershaw-brown/DART/tree/scaleh
